lib/Exporter/ONNXModelWriter.cpp

```diff
@@ -0,0 +1,1367 @@
+}
+
+void ONNXModelWriter::tensorShapeFromPlaceholder(const Placeholder *PH,
+                                                 ValueInfoType *valueProto) {
+  valueProto->set_name(PH->getName());
+  auto *type = valueProto->mutable_type();
+  auto *tensorType = type->mutable_tensor_type();
+  auto *glowType = PH->getType();
+  tensorType->set_elem_type(convertType(*glowType));
+  auto *tensorShape = tensorType->mutable_shape();
+  const auto &dims = glowType->dims();
+  for (unsigned b = 0, e = dims.size(); b < e; ++b) {
+    auto *tensorDims = tensorShape->add_dim();
+    tensorDims->set_dim_value(dims[b]);
+  }
+}
+
+bool ONNXModelWriter::isIgnoreNode(const Node *node) {
+  switch (node->getKind()) {
+  case Kinded::Kind::SplatNodeKind:
+  case Kinded::Kind::TileNodeKind:
+    return true;
+  default:
+    return false;
+  }
+}
+
+const Node *ONNXModelWriter::unwindBroadcastInput(const Node *tile,
+                                                  ReporterType &reporter) {
+  while (tile && tile->getKind() == Kinded::Kind::TileNodeKind) {
+    reporter(tile);
+    tile = tile->getNthInput(TileNode::InputIdx).getNode();
+  }
+  return tile;
+}
+
+void ONNXModelWriter::outputsToProto(const Node *node, NodeType *proto) {
+  for (const auto &use : node->getUsers()) {
+    const auto *user = use.getUser();
+    if (user->getKind() == Kinded::Kind::SaveNodeKind) {
+      proto->add_output(user->getName());
+    } else {
+      // Find the user input that matches input node parameter
+      for (unsigned b = 0, e = user->getNumInputs(); b < e; ++b) {
+        auto *UIN = user->getNthInput(b).getNode();
+        if (UIN == node) {
+          proto->add_output(UIN->getName());
+          break;
+        }
+      }
+    }
+  }
+}
+
+void ONNXModelWriter::inputsToProto(const Node *node, NodeType *proto) {
+  for (unsigned b = 0, e = node->getNumInputs(); b < e; ++b) {
+    proto->add_input(node->getNthInput(b).getNode()->getName());
+  }
+}
+
+llvm::Error ONNXModelWriter::writeAllWithNode(const std::string &opName,
+                                              const Node *node,
+                                              NodeType *proto) {
+  proto->set_name(node->getName());
+  proto->set_op_type(opName);
+  inputsToProto(node, proto);
+  outputsToProto(node, proto);
+  return llvm::Error::success();
+}
+
+llvm::Error ONNXModelWriter::writeAll(const std::string &opName,
+                                      const Node *node, GraphType &graph) {
+  return writeAllWithNode(opName, node, graph.add_node());
+}
+
+llvm::Error ONNXModelWriter::writeArithmetic(const std::string &opName,
+                                             const Node *node, GraphType &graph,
+                                             ReporterType &reporter) {
+  auto *proto = graph.add_node();
+  proto->set_name(node->getName());
+  proto->set_op_type(opName);
+  outputsToProto(node, proto);
+
+  const auto *LHS = node->getNthInput(MulNode::LHSIdx).getNode();
```

Comment:
You may want to explain why `MulNode::LHSIdx` is used here ;-) I guess because all arithmetic binary ops are structurally the same? May be introduce a helper var like: `unsigned lhsIdx = MulNode::LHSId;` and use it instead. Same for the RHS.